kore/src/Kore/Rewrite/SMT/Evaluator.hs

```diff
@@ -244,39 +248,83 @@ translateTerm smtType (QuantifiedVariable var) = do
     return smtVar
 translateTerm t (UninterpretedTerm (TermLike.ElemVar_ var)) =
     lookupVariable var <|> declareVariable t var
+translateTerm t (UninterpretedPredicate predicate) = do
+    TranslatorState{quantifiedVars, predicates} <- State.get
+    let freeVars = FreeVariables.freeVariableNames @_ @variable predicate
+        boundVarsMap = filterBoundVarsMap freeVars quantifiedVars
+        boundPat =
+            Predicate.makeExistsPredicateN (Map.keys boundVarsMap) predicate
+        conversion = Predicate.fromPredicate (mkSortVariable "_")
+        stateSetter = field @"predicates"
+    lookupUninterpreted boundPat quantifiedVars predicates
+        <|> declareUninterpreted t conversion stateSetter boundPat boundVarsMap
 translateTerm t (UninterpretedTerm pat) = do
     TranslatorState{quantifiedVars, terms} <- State.get
-    let freeVars =
-            TermLike.freeVariables @_ @variable pat
-                & FreeVariables.toNames
-        boundVarsMap =
-            Map.filterWithKey
-                (\k _ -> inject (variableName k) `Set.member` freeVars)
-                quantifiedVars
+    let freeVars = FreeVariables.freeVariableNames @_ @variable pat
+        boundVarsMap = filterBoundVarsMap freeVars quantifiedVars
         boundPat = TermLike.mkExistsN (Map.keys boundVarsMap) pat
+        stateSetter = field @"terms"
     lookupUninterpreted boundPat quantifiedVars terms
-        <|> declareUninterpreted boundPat boundVarsMap
-  where
-    lookupUninterpreted boundPat quantifiedVars terms =
-        maybe empty (translateSMTDependentAtom quantifiedVars) $
-            Map.lookup boundPat terms
-    declareUninterpreted boundPat boundVarsMap =
+        <|> declareUninterpreted t id stateSetter boundPat boundVarsMap
+
+declareUninterpreted ::
+    (MonadSMT m, MonadLog m, InternalVariable variable, Ord termOrPredicate) =>
+    SExpr ->
+    (termOrPredicate -> TermLike variable) ->
+    Lens.ASetter
+        (TranslatorState variable)
+        (TranslatorState variable)
+        (Map.Map termOrPredicate (SMTDependentAtom variable))
+        (Map.Map termOrPredicate (SMTDependentAtom variable)) ->
+    termOrPredicate ->
+    Map.Map (ElementVariable variable) (SMTDependentAtom variable) ->
+    Translator variable m SExpr
+declareUninterpreted
+    sExpr
+    termLikeConversion
+    stateSetter
+    boundPat
+    boundVarsMap =
         do
             n <- Counter.increment
-            logVariableAssignment n boundPat
+            logVariableAssignment n (termLikeConversion boundPat)
```

Comment:
Is it necessary for `logVariableAssignment` to take a `TermLike`? Would taking some `t` where `t` has a `Pretty` instance work?